packages/remix-react/browser.tsx

```diff
@@ -177,6 +178,21 @@ export function RemixBrowser(_props: RemixBrowserProps): ReactElement {
           window.__remixContext.future.v2_normalizeFormMethod,
       },
     });
+
+    // Hard reload if the URL we tried to load is not the current URL.
+    // This is usually the result of 2 rapid backwards/forward clicks from an
+    // external site into a Remix app, where we initially start the load for
+    // one URL and while the JS chunks are loading a second forward click moves
+    // us to a new URL
+    let initialUrl = window.__remixContext.url;
+    let hydratedUrl = window.location.pathname + window.location.search;
```

Comment:
Minor suggestion that this could be formalized into a fn which is used here and in server.ts to construct the string for matching. Would prevent a scenario where a future engineer sees `url` and assumes they could simply `window.location.toString()` this instead of relying only on the path and search params.

Reply:
We can't share util functions like that across the package boundaries so we wouldn't be able to DRY it up much.  I think it' ok for now since I don't expect this to change much - and we're pretty familiar with the fact that on the server a URL is only ever `pathname` and `search`, never `hash` 👍